service/history/workflowTaskHandlerCallbacks.go

```diff
@@ -564,7 +557,16 @@ Update_History_Loop:
 
 		if workflowTaskHeartbeatTimeout {
 			// at this point, update is successful, but we still return an error to client so that the worker will give up this workflow
-			return nil, serviceerror.NewNotFound(fmt.Sprintf("workflow task heartbeat timeout"))
+			return nil, serviceerror.NewNotFound("workflow task heartbeat timeout")
+		}
+
+		if workflowTaskFailedErr != nil {
+			switch workflowTaskFailedErr.causeErr.(type) {
+			case nil:
+				return nil, serviceerror.NewInternal(workflowTaskFailedErr.Error())
+			case *serviceerror.InvalidArgument:
+				return nil, serviceerror.NewInvalidArgument(workflowTaskFailedErr.Error())
+			}
```

Comment:
This is an actual change of this PR. I am not sure about `nil` case though. It only happens when `handler.hasBufferedEvents` is `true` and reason is `WORKFLOW_TASK_FAILED_CAUSE_UNHANDLED_COMMAND`. Should it be `InvalidArgument` or `Internal` fits better here?

Reply:
I don't think we should be setting the workflowTaskError to nil when we return an error back to user because workflow task has Completed command and there are some unhandled events in the buffer.  I think we should be returning back InvalidArgument error back to the caller because we don't want user to retry this workflow task completion.